README.md
Gradient
========


ChangeLog.txt
0.2.1
findContours + approxPolyDP
  Added: prints misaligned blended image if some questions were missing
  Added: refactored Main module, grades after 6 or so refinements or after pressing escape
  Fixed: findPoints relies entirely on findcontours and approxPolyDP from opencv to find the corner points

0.1.1
AnswerLayout Refactor
  Added: reliance on numpy arrays for all point arithmetic
  Added: class Box representing boxes drawn on the answer sheets
  Fixed: answerMap coordinates are now percentages, and all code depending on this structure was adjusted accordingly

0.1.0
Answer Sheet Generation
  Added: AnswerLayout module heeding Obama's call to reduce dependency on foreign answer sheets
          -produces answerMap structure which is based on coordinates depending on the image dimensions (pageDims)
          -generates randomly bubbled answer sheets for testing
  Fixed: Main, Grade, Transformation modules depend on answer sheets generated by AnswerLayout
  Remvd: thresholding of registered image. currently kmeans on convolutions give good split for distinguishing shading with grayscale

0.0.4
Registration Validation
  Added: minimizing sum(bitwise_and(registered, template)) as criterion for optimal registration
  Added: quota for the number of improvements to the registration before grading takes place
  Added: function getAnswerMap returns structure of answer coords according to the specification in google docs

0.0.3
Perspective Transform and Canny Refinement
  Added: Main - View variable for debugging purposes, shows a set of images based on its value
  Added: Main - Suppresion variable to suppress errors being printed to console
  Added: Preprocessing - Morphological tranformations to the canny output
  Added: FindPoints - now outputs 4 points, as needed for the perspective transform
  Fixed: Transform - transformation is now a 3D perspective transform, fourth point of template hard coded in source
  Added: Grade - kmeans on convolution results to find threshold, currently crudely taking average between high and low
  Added: Grade - deprecated code based on ideal template (425, 550) metrics


0.0.2
Modularization
  Added: Corner validation now mostly works (methods added to FindPoints.py: findCorners(), partitionPoints())
  Added: Code refactor, compartmentalized algorithm into separate modules.
          The Main file is the main loop that calls the following
          Preprocess.py accepts the image, the new shape, and the big and small square regions.
            It returns the edges image with only the region drawn and the original image with the square
            regions drawn and resized.
          FindPoints.py accepts the edges image and the image.
            It returns the list of intersections and the image with the intersections drawn on.
          Transform.py accepts the image and the intersection points used to transform.
            It returns the registered image and the blend of the image with the template.
          Grade.py accepts the registered.
            It does not return but will list out the answers.
  Added: Error.py to contain the user defined errors that are now being raised

0.0.1
CornerValidate:
  FIXED: houghlines had an improper parameter set
  Added: validateCorners() (incomplete)
  Added: offset() method to determine the offset values based on the window size
  Added: global variables pi and win as the number pi and the window size
  Added: drawSquare() to draw a square on the image, used to help user align the sheet
  Added: region() to black out all except the expected area for the answer sheet to be located
  Added: Check to make sure at least 4 lines are detected
  Added: filter the result of kmeans() on the detected lines so only lines with certain angles show (may not work)
  added: Check to make sure at least 4 intersections are found
  Added: run kmeans() on the intersection points to obtain only 4 points


Main.py
import cv2
import numpy as np
from scipy.optimize import minimize
from Preprocess import win, draw_square, preprocess, region, show_intersections
from FindPoints import findPoints
from Transform import transform, myDims
from Grade import grade, ndrawAnswerCoords
import Errors

# debugging
import sys
import traceback
import time

camera_port = 0
ramp_frames = 4

suppress = True

# debugging
debug = \
    {
        'suppress': True
    }

views = \
    {
        'reg': True,
        'corners': False,
        'edges': False,
        'thresh': False,
        'blend': True,
        'visual': False
    }


# to turn on/off views at runtime, view list above is global
def toggle_view(view_name, image):
    if views[view_name]:
        cv2.imshow(view_name, image)
    elif cv2.getWindowProperty(view_name, cv2.CV_WINDOW_AUTOSIZE) > 0:  #  get any window prop to check for existence
        cv2.destroyWindow(view_name)


# makes a function to pass to minimization
def affine_dec(image, template_math):
    rows, cols = image.shape

    def affine(d):
        [dx, dy] = d
        m = np.float32([[1, 0, dx], [0, 1, dy]])
        dst = cv2.warpAffine(image, m, (cols, rows))
        return np.sum(cv2.bitwise_and(dst, template_math))

    return affine


def main():
    # for criterion, minval can be any exorbitantly large value
    minval = 500000000
    quota = 0
    
    camera = cv2.VideoCapture(camera_port)

    for i in xrange(ramp_frames):
        retval, temp = camera.read()


    while True:
        start_time = time.time()
        # The function waitKey waits for a key event infinitely (when delay<=0 )
        # or for delay milliseconds, when it is positive. It returns the code of
        # the pressed key or -1 if no key was pressed before the specified time
        # had elapsed. Escape code is 27
        wk = cv2.cv.WaitKey(10)
        print wk

        # registration criterion, only grade if we've improved 7 times
        # break on space bar
        if quota == 20 or wk == 32:
            # if registered exists, try grading it. otherwise, break
            # if reg exists and you grade it successfully, break. o/w keep going
            if 'registered' in locals():
                try:
                    print 'grading...'
                    if not grade(registered):
                        cv2.imwrite('misaligned.jpg', blend_visual)
                        quota = 0
                    else:  # we stop only when we get a complete grade
                        break
                except:
                    if not suppress:
                        print "something bad is happening at Grade"
            else:
                break

        # escapes
        if wk == 27:  #  escape
            break

        # dynamic views, there's a better structure for this
        if wk == 114:  # r
            views['reg'] = not views['reg']
        if wk == 99:  # c
            views['corners'] = not views['corners']
        if wk == 101:  # e
            views['edges'] = not views['edges']
        if wk == 116:  # t
            views['thresh'] = not views['thresh']
        if wk == 98:  # b
            views['blend'] = not views['blend']
        if wk == 118:  # v
            views['visual'] = not views['visual']

        # output control
        if wk == 115:
            debug['suppress'] = not debug['suppress']

        retval, im = camera.read()

        try:
            im = np.array(im[:, :im.shape[0], :])
            im = cv2.resize(im, win)

            # present feed for input
            draw_square(im)
            cv2.imshow('input', cv2.flip(im, 1))

            gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)
            edges = cv2.Canny(gray, 100, 240)
            region(edges, margin=0)

            toggle_view('edges', edges)
            
            intersections = findPoints(edges)

        except Errors.ImproperIntersectionsError as e:
            if not suppress:
                print e
            continue
        except Exception as e:
            if not suppress:
                print e, traceback.format_exception(*sys.exc_info())
            continue
        
        # if no points found, this code isn't seen
        # show computed intersection points
        for point in intersections:
            cv2.circle(im, (int(round(point[0])), int(round(point[1]))), 6, (0, 0, 255), -1)
        toggle_view('corners', cv2.flip(im, 1))

        try:
            registered = transform(im, intersections)
            toggle_view('reg', registered)
        except Errors.NotEnoughPointsToTransformError as e:
            if not suppress:
                print e
            continue
        
        # adjust registered
        registered = cv2.cvtColor(registered, cv2.COLOR_BGR2GRAY)
        retval, registered_bin = cv2.threshold(registered, 120, 255, cv2.THRESH_BINARY+cv2.THRESH_OTSU)

        '''==========BLENDING AND REGISTRATION CRITERION=========='''
                    # -----------------------------
                    # This is for visual comparison
                    # -----------------------------
        if views['visual']:
            # template to show difference between it and the registered image
            # this read incurs a 5-fold decrease in speed
            template_visual = cv2.cvtColor(cv2.imread('blank.jpg'), cv2.COLOR_BGR2GRAY)
            # reduce to dimensions to match registered
            template_visual = cv2.resize(template_visual, (0, 0), fx=0.25, fy=0.25)
            blend_visual = cv2.addWeighted(registered, .7, template_visual, .3, 0)
            cv2.imshow('visual blend', blend_visual)
        elif cv2.getWindowProperty('visual blend', cv2.CV_WINDOW_AUTOSIZE) > 0:  #  get any window prop to check for existence
                cv2.destroyWindow('visual blend')
                    # -----------------------------------
                    # This is for mathematical comparison
                    # -----------------------------------
        # create criterion template, reverse size tuple
        template_math = np.zeros(myDims[::-1], dtype=np.uint8)
        template_math = ndrawAnswerCoords(template_math)

        # compute blend
        blend_math = cv2.bitwise_and(registered_bin, template_math)

        # impose criterion
        blend_sum = np.sum(blend_math)
        toggle_view('blend', blend_math)
        if blend_sum < minval:
            # take {quote} improvements before showing grade
            print blend_sum
            quota += 1
            minval = blend_sum

        print "time: ", time.time()-start_time
        '''======================================================='''
        
    cv2.cv.DestroyAllWindows()
    camera.release()

if __name__ == '__main__':
    main()


Grade.py
import cv2
import numpy as np
from AnswerLayout import answerMap, pageDims

#IMPORTANT. radius should be adjusted programmatically, not manually
radius = 10

def ndrawAnswerCoords(im):
    #shape is given as height by width, so it needs to be reversed
    imDims = np.array(list(im.shape[::-1]))
    for num in answerMap:
        for letter in answerMap[num]:
            loc = tuple(np.round(imDims*answerMap[num][letter]).astype(int))
            cv2.circle(im, loc , radius/2, (255,255,255), -1)
    return im
    
def convolve(im, point):
    sum = 0
    for i in range(0, 2 * radius):
        for j in range(0, 2 * radius):
            sum += im[point[1] - radius + i][point[0] - radius + j]
    return sum

'''the registered image is mutated'''
def grade(registered):
    
    #adjust image to match the coords from answerMap
    tempSize = registered.shape
    registered = cv2.resize(registered, tuple(pageDims[::-1]))
    imDims = np.array(list(registered.shape[::-1]))

    #generate local answerMap
    local_answerMap = {}
    for num in answerMap:
        for letter in answerMap[num]:
            loc = tuple(np.round(imDims*answerMap[num][letter]).astype(int))
            if not num in local_answerMap.keys():
                local_answerMap[num] = {}
            local_answerMap[num][letter] = loc
        
    #thresholding
    retval, registered = cv2.threshold(registered, 80, 255, cv2.THRESH_BINARY_INV)


    #there is actually a perfect morphological opening based on circle radius
    #that would remove noise
    registered = cv2.erode(registered, np.ones((3,3)), iterations=6)
    registered = cv2.dilate(registered, np.ones((3,3)), iterations=6)

    #for every number, check which letter has highest convolution
    answers = {}
    for num in local_answerMap:
        max = 0
        for letter in local_answerMap[num]:
            res = convolve(registered, local_answerMap[num][letter])
            if res>max:
                max = res
                answers[num] = letter

    #print answers and check if any were missing
    count = 0
    missing = []
    for num in sorted(answers.iterkeys()):
        print "%d: %c" % (num, answers[num])
        if not count == num:
            missing.append(count)
            count += 1
            while not count == num:
                missing.append(count)
                count += 1
            count += 1
        else:
            count += 1

    #not sure why this is being resized here
    registered = cv2.resize(registered, tempSize[::-1])

    #if numbers are missing, print the registered image
    if len(missing) == 0:
        print "complete grade"
        return True
    else:
        print "missing: ", missing
        return False


Transform.py
import cv2
import numpy as np
import Errors
from AnswerLayout import container, pageDims

'''template dims hard coded'''
# pageDims are (width, height)
myDims = [530, 750]


def transform(im, points):
    # four points and their mapped versions used to compute underlying perspective transform
    # lexicographically ordered list
    p_list = [np.round((myDims * p)/pageDims).astype(int).tolist() for p in container.p_list]

    # error if points are float32
    dst_square = np.array(p_list, np.float32)
    src_square = np.array(points, np.float32)
    
    if not (len(dst_square) - len(src_square)) == 0:
        raise Errors.NotEnoughPointsToTransformError
    
    # compute transform based on point mappings above
    _transform = cv2.getPerspectiveTransform(src_square, dst_square)

    registered = cv2.warpPerspective(im, _transform, tuple(myDims))

    return registered


FindPoints.py
import cv2
import numpy as np
import math
from scipy.cluster.vq import kmeans, whiten
import Errors

pi = cv2.cv.CV_PI

def lineToPointPair(line):

    r = line[0]
    t = line[1]
    cost = math.cos(t)
    sint = math.sin(t)
    x0 = r*cost
    y0 = r*sint
    alpha = 1000

    points = []
    
    points.append([x0+alpha*(-sint), y0+alpha*cost])
    points.append([x0-alpha*(-sint), y0-alpha*cost])

    return points

def computeIntersect(line1, line2):
    [l1p1, l1p2] = lineToPointPair(line1)
    [l2p1, l2p2] = lineToPointPair(line2)
    [x1, y1] = l1p1
    [x2, y2] = l1p2
    [x3, y3] = l2p1
    [x4, y4] = l2p2

    denom = (x1 - x2) * (y3-y4) - (y1 - y2) * (x3 - x4)
    intersect = [(((x1 * y2) - (y1 * x2)) * (x3 - x4)
                  - (x1 - x2) * ((x3 * y4) - (y3 * x4))) / denom,
                 (((x1 * y2) - (y1 * x2)) * (y3 - y4)
                  - (y1 - y2) * ((x3 * y4) - (y3 * x4))) / denom]

    return intersect

def acceptLinePair(line1, line2):
    minTheta = (cv2.cv.CV_PI) / 32
    minThetaDiff = 3 * (cv2.cv.CV_PI) / 8
    maxThetaDiff = 5 * (cv2.cv.CV_PI) / 8
    theta1 = line1[1]
    theta2 = line2[1]

    if (theta1 < minTheta):
        theta1 += pi #deals with 0 and 180 ambiguity
    if (theta2 < minTheta):
        theta2 += pi #deals with 0 and 180 ambiguity

    thetaDiff = abs(theta1-theta2)
    return thetaDiff > minThetaDiff and thetaDiff < maxThetaDiff

def drawIntersection(detectedLines, image):
    #compute the intersection from the lines detected...
    intersections = []
    for i in range(0, detectedLines.shape[0]):
        for j in range(i, detectedLines.shape[0]):
            line1 = detectedLines[i]
            line2 = detectedLines[j]
            if(acceptLinePair(line1, line2)):
                intersection = computeIntersect(line1, line2)
                intersections.append(intersection)
                
    #draw intersections on given image   
    if len(intersections) > 0:
        for i in range(0, len(intersections)):
            p_i = (int(round(intersections[i][0])), int(round(intersections[i][1])))
            cv2.circle(image, p_i , 1, (0, 255, 0), 3)
    return intersections

#ideas: compute the center mass from the points and then we can separate the points
    #by those that are above and below then by left and right. if we don't have 4 separate
    # points then it failed. if we do then we have to check further. (how?)
def validateCorners(points):
    above, below, left, right = partitionPoints(points)
    if not (len(above) == 2 and len(below) == 2
            and len(left) == 2 and len(right) == 2):
        return False
    
    return True

def findCorners(points):
    above, below, _, _ = partitionPoints(points)
    topL = min(above, key = lambda (p1,p2): p1+p2)
    topR = max(above, key = lambda (p1,p2): p1+p2)
    botL = min(below, key = lambda (p1,p2): p1+p2)
    botR = max(below, key = lambda (p1,p2): p1+p2)

    return topL, topR, botL, botR

def partitionPoints(points):
    cx = sum(points[:][0]) / len(points)
    cy = sum(points[:][1]) / len(points)
    above = [p for p in points if p[1] < cy]
    below = [p for p in points if p[1] > cy]
    left = [p for p in points if p[0] < cx]
    right = [p for p in points if p[0] > cx]

    return above, below, left, right

def findPoints(edges, im):
    try:
        #flatten list one level
        lines = cv2.HoughLines( edges, 1, cv2.cv.CV_PI/180, 100)[0]
    except TypeError:
        raise Errors.NotEnoughCannyEdgesError

    if len(lines) < 4:
        raise Errors.NotEnoughCannyEdgesError

    lines = kmeans(lines,8)[0]

    width = pi / 8
    lines = np.array([l for l in lines
                      if not ((l[1] > (pi / 4) - width and l[1] < (pi / 4) + width)
                      or (l[1] > (3 * pi / 4) - width and l[1] < (3 * pi / 4) + width))])

    for line in lines:
        p1 = (int(round(lineToPointPair(line)[0][1])), int(round(lineToPointPair(line)[0][1])))
        p2 = (int(round(lineToPointPair(line)[1][0])), int(round(lineToPointPair(line)[1][1])))
        cv2.line(edges, p1, p2, (255,255,255))

    #compute intersections points of the resultant lines
    intersections = drawIntersection(lines, im)
    
    if len(intersections) < 4:
        raise Errors.NotEnoughIntersectionsError

    intersections = kmeans(np.array(intersections), 4)[0].tolist()

    if not validateCorners(intersections):
        raise Errors.ImproperIntersectionsError

    topL, topR, botL, botR = findCorners(intersections)

    return [topL, topR, botL, botR], im


Preprocess.py
import cv2
import numpy as np

big = ((80,80), (420,420))
small = ((150,150), (350,350))

win = (500, 500)
r_big = 180
# division between integers produces integer (by rounding)
big = (((win[0]/2)-r_big, (win[1]/2)-r_big), ((win[0]/2)+r_big, (win[1]/2)+r_big))


def draw_square(im):
    ((xo, yo), (x, y)) = big
    cv2.line(im, (xo,yo), (x,yo), (0,255,0))
    cv2.line(im, (x,yo), (x,y), (0,255,0))
    cv2.line(im, (x,y), (xo,y), (0,255,0))
    cv2.line(im, (xo,y), (xo,yo), (0,255,0))


def region(im, margin):
    im[:big[0][0]-margin, :] = 0 #top
    im[:, :big[0][1]-margin] = 0 #left
    im[big[1][0]+margin:im.shape[1], :] = 0 #bottom
    im[:, big[1][1]+margin:im.shape[0]] = 0 #right


def show_intersections(im, intersections):
    for point in intersections:
        cv2.circle(im, (int(round(point[0])), int(round(point[1]))),
                   6, (0, 0, 255), -1)
    cv2.imshow("corners", im)


def preprocess(im):
    #why do we cut the image again?
    im = np.array(im[:,:im.shape[0],:])
    im = cv2.resize(im, win)
    gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)
    
    edges = cv2.Canny(gray, 100, 240)

    drawSquare(im, big)
    drawSquare(im, small)
    
    region(edges)

    return edges, im


core.py
import cv2
import math
import numpy as np
from scipy.cluster.vq import kmeans, whiten

#425, 550

pi = cv2.cv.CV_PI
win = (500,500)

def drawCircles(im, squareOrigin):
    (ox, oy) = (squareOrigin[0], squareOrigin[1])
    (h, w) = (im.shape[0], im.shape[1]) #first two dimensions w and h

    bubbles = []

    radius = int(round(4*(0.009411*h)))
    rows = 17
    col = 2
    for i in range(0,rows):
        wOffset = 0.070588*w
        hOffset = 0.07636*h
        circlex = ox+wOffset
        circley = oy+hOffset+i*(0.02625*h)
        for j in range(0,5):
            circle_1 = (int(round(circlex)), int(round(circley)))
            circle_2 = (int(round(circlex+(0.18824*w))), int(round(circley)))
            #cv2.circle(im, circle_1 , 5, (255,255,255), -1)
            #cv2.circle(im, circle_2 , 5, (255,255,255), -1)
            bubbles+=[circle_1,circle_2]
            circlex+=(0.025882*w)

    return bubbles

def convolve(im, point):
    #radius should not be hardcoded
    radius = 4
    sum = 0
    for i in range(0,2*radius):
        for j in range(0,2*radius):
            sum+=im[point[1]-radius+i][point[0]+-radius+j]
    return sum

def offset():
    w = 0.070588 / 425 * win[1]
    h = 0.07636 / 550 * win[0]
    h2 = 0.02625 / 550  * win[0]
    w2 = 0.025882 / 425 * win[1]
    return (w,h, h2, w2)

def findAnswer(im, squareOrigin, (x,y)):
    #given a pair of coordinates (x,y), finds (question number, letter answered)
    (ox, oy) = (squareOrigin[0], squareOrigin[1])
    (h, w) = im.shape
    wOffset, hOffset, h2, w2 = offset()
    wOffset *= w
    hOffset *= h

    rows = 17
    col = 2
    
    j = (y - (oy+hOffset))/(h2*h)
    if int((x-(ox+wOffset))/70) == 0:
        i = (x-(ox+wOffset))/(w2*w)
    else:
        i = ((x-(ox+wOffset+80))/(w2*w))+5
        j+=rows

    return (int(round(j)),int(round(i)))
    
def acceptLinePair(line1, line2):
    minTheta = (cv2.cv.CV_PI)/32
    minThetaDiff = 3*(cv2.cv.CV_PI)/8
    maxThetaDiff = 5*(cv2.cv.CV_PI)/8
    theta1 = line1[1]
    theta2 = line2[1]

    if (theta1 < minTheta):
        theta1 += cv2.cv.CV_PI #deals with 0 and 180 ambiguity
    if (theta2 < minTheta):
        theta2 += cv2.cv.CV_PI #deals with 0 and 180 ambiguity

    thetaDiff = abs(theta1-theta2)
    return thetaDiff > minThetaDiff and thetaDiff < maxThetaDiff

def computeIntersect(line1, line2):
    [l1p1, l1p2] = lineToPointPair(line1)
    [l2p1, l2p2] = lineToPointPair(line2)
    [x1,y1] = l1p1
    [x2,y2] = l1p2
    [x3,y3] = l2p1
    [x4,y4] = l2p2

    denom = (x1-x2)*(y3-y4) - (y1-y2)*(x3-x4)
    intersect = [((x1*y2-y1*x2)*(x3-x4)
               -(x1-x2)*(x3*y4-y3*x4))/denom,
               ((x1*y2-y1*x2)*(y3-y4)
                -(y1-y2)*(x3*y4-y3*x4))/denom]

    return intersect

def lineToPointPair(line):

    r = line[0]
    t = line[1]
    cost = math.cos(t)
    sint = math.sin(t)
    x0 = r*cost
    y0 = r*sint
    alpha = 1000

    points = []
    
    points.append([x0+alpha*(-sint), y0+alpha*cost])
    points.append([x0-alpha*(-sint), y0-alpha*cost])

    return points

def drawIntersection(detectedLines, image):
    #compute the intersection from the lines detected...
        intersections = []
        for i in range(0,detectedLines.shape[0]):
            for j in range(i,detectedLines.shape[0]):
                line1 = detectedLines[i]
                line2 = detectedLines[j]
                if(acceptLinePair(line1, line2)):
                    intersection = computeIntersect(line1,line2)
                    intersections.append(intersection)
        #draw intersections on given image   
        if len(intersections)>0:
            for i in range(0,len(intersections)):
                p_i = (int(round(intersections[i][0])),int(round(intersections[i][1])))
                cv2.circle(image, p_i , 1, (0,255,0), 3)
        return intersections

#ideas: compute the center mass from the points and then we can separate the points
    #by those that are above and below then by left and right. if we don't have 4 separate
    # points then it failed. if we do then we have to check further. (how?)
def validateCorners(points):
    '''
    for p in points:
        print p
    pmin = min(points, key = lambda (a,b): a + b)
    pmax = max(points, key = lambda (a,b): a + b)
    points = [p for p in points if not (p == pmin or p == pmax)]
    dist_min1 = math.hypot(points[0][0] - pmin[0], points[0][1] - pmin[1])
    dist_min2 = math.hypot(points[1][0] - pmin[0], points[1][1] - pmin[1])
    dist_max1 = math.hypot(points[0][0] - pmax[0], points[0][1] - pmax[1])
    dist_max2 = math.hypot(points[1][0] - pmax[0], points[1][1] - pmax[1])
    ''''''
    dist1 = math.hypot(p2[0] - p1[0], p2[1] - p1[1])
    dist2 = math.hypot(p3[0] - p1[0], p3[1] - p1[1])
    dist3 = math.hypot(p4[0] - p1[0], p4[1] - p1[1])
    longest = max([dist1, dist2, dist3])
    ''''''
    print pmin
    print pmax
    '''
    return True

#==================================
#==================================
'''NEXT STEPS'''
# 1. restrict attention to the ROI within the white paper (this should be easy)
# 2. figure out the issue behind FLAG1

#==================================
#==================================

def drawSquare(im, ((xo, yo), (x, y))):
    cv2.line(im, (xo,yo), (x,yo), (0,255,0))
    cv2.line(im, (x,yo), (x,y), (0,255,0))
    cv2.line(im, (x,y), (xo,y), (0,255,0))
    cv2.line(im, (xo,y), (xo,yo), (0,255,0))

def region(im, big, small):
    im[:big[0][0],:] = 0
    im[:,:big[0][1]] = 0
    im[big[1][0]:im.shape[1],:] = 0
    im[:,big[1][1]:im.shape[0]] = 0
    im[small[0][0]:small[1][0],small[0][1]:small[1][1]] = 0

def PreProcessing(im):
    sbig = ((80,80), (420,420))
    ssmall = ((150,150), (350,350))
    temp = im
    im = np.array(im[:,:im.shape[0],:])
    im = cv2.resize(im, win)
    gray = cv2.cvtColor(im,cv2.COLOR_BGR2GRAY)
    reval, thresh = cv2.threshold(gray, 200.0, 255.0, cv2.THRESH_BINARY_INV | cv2.THRESH_OTSU)
    
    drawSquare(im, sbig)
    drawSquare(im, ssmall)

    edges = cv2.Canny(gray, 100.0, 240.0)

    region(edges, sbig, ssmall)
    cv2.imshow('edges', edges)
    return edges

def FindPoints(edges):
    
    try:
        lines = cv2.HoughLines( edges, 1, cv2.cv.CV_PI/180, 100)[0] #flatten list one level
    except TypeError:
        print "Not enough edges detected by canny"
        return []
    
    if len(lines) < 4:
        return []
    #print "Detected "+str(lines.shape[0])+" lines"

    lines = kmeans(lines,8)[0]
    width = pi / 8
    lines = np.array([l for l in lines
                      if not ((l[1] > (pi / 4) - width and l[1] < (pi / 4) + width)
                      or (l[1] > (3 * pi / 4) - width and l[1] < (3 * pi / 4) + width))])
    
    for line in lines:
        p1 = (int(round(lineToPointPair(line)[0][0])),int(round(lineToPointPair(line)[0][1])))
        p2 = (int(round(lineToPointPair(line)[1][0])),int(round(lineToPointPair(line)[1][1])))
        cv2.line(edges, p1, p2, (255,255,255))

    #compute intersections points of the resultant lines, dropping right-most, bottom-most point
    intersections = drawIntersection(lines, im)
    if len(intersections) < 4:
        return
    #intersections = whiten(np.array(intersections))
    intersections = kmeans(np.array(intersections), 4)[0].tolist()

    if not validateCorners(intersections):
        return

    if (intersections == None):
        print "none"
        return []
    if (len(intersections)==0):
        print 0
        return []

    #show computed intersection points
    try:
        for point in intersections:
            cv2.circle(im, (int(round(point[0])),int(round(point[1]))) , 6, (0,0,255), -1)
    except OverflowError:
        return []
    
    cv2.imshow('dfdf', im)
    
    intersections.sort()
    intersections.pop()

    cv2.imshow('dfja', im)

    return intersections

def Transform(im, points):
    #three points and their mapped versions used to compute underlying affine transform
    dstTriangle = np.array([[25,45],[25,336],[307,45]], np.float32)
    srcTriangle = np.array(points,np.float32)
    
    if(not (len(dstTriangle)-len(srcTriangle))==0):
        return
    
    #compute transform based on point mappings above
    transform = cv2.getAffineTransform(srcTriangle,dstTriangle)

    #get image dimension (redundant), then apply derived transform to obtain registered image
    rows,cols,depth = im.shape
    registered = cv2.warpAffine(im, transform, (cols,rows))
    
    #adjust template to show difference between it and the registered image
    template = cv2.cvtColor(cv2.imread('answerTemplate.jpg'),cv2.COLOR_BGR2GRAY)
    blank = np.ones((2200, 2200), dtype=np.uint8) * 255
    blank[:,:1700] = template
    #template = cv2.resize(blank, (500,500))
    template = cv2.resize(blank, (640, 480))

    print template.shape
    print registered.shape
    registered = cv2.cvtColor(registered, cv2.COLOR_BGR2GRAY)    
    blend = cv2.addWeighted(registered, .7, template, .3, 0)
    
    cv2.imshow("blend", blend)

    return registered

def Grade(registered):
    #cv2.imshow("warped.png",registered)
    #doesn't actually draw circles, returns locations of bubble centers taken from perfect template
    #the ordered pair argument is the offset to the upperleft corner, shouldn't be hardcoded
    bubbles = drawCircles(registered, (29,49))

    #for every bubble coordinate, convolve that location with a kernal to see if that answer was bubbled
    answers = {}
    '''i should fix threshold for convolution'''
    for pair in bubbles:
        if(convolve(registered, pair)>7000):
            (num, ans)=findAnswer(registered, (25,45), pair)
            letter = "A"
            if(ans==0 or ans==5):
                letter = "A"
            elif(ans==1 or ans==6):
                letter = "B"
            elif(ans==2 or ans==7):
                letter = "C"
            elif(ans==3 or ans==8):
                letter = "D"
            elif(ans==4 or ans==9):
                letter = "E"
            answers[num+1]=letter
            #print "ans: "+str(num+1)+", "+letter
    for key in sorted(answers.iterkeys()):
        print "%d: %s" %(key, answers[key])
    
camera_port = 0
ramp_frames = 4
camera = cv2.VideoCapture(camera_port)
def get_image():
    retval, im = camera.read()
    return im

for i in xrange(ramp_frames):
    temp = get_image()
    print temp

while True:
    if cv2.cv.WaitKey(10) == 27:
        break
    im = get_image()

    edges = PreProcessing(im)
    intersections = FindPoints(edges)
    if(intersections == None):
        continue
    if(len(intersections)==0):
        print "Error finding points"
        continue
    registered = Transform(im, intersections)
    if (registered == None):
        continue
    Grade(registered)
cv2.cv.DestroyAllWindows()
